src.freebsd/coreutils/date/date.1
.\"-
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd November 10, 2025
.Dt DATE 1
.Os
.Sh NAME
.Nm date
.Nd display or set date and time
.Sh SYNOPSIS
.\" Display time.
.Nm
.Op Fl nRu
.Op Fl z Ar output_zone
.Op Fl I Ns Op Ar FMT
.Op Fl r Ar filename
.Op Fl r Ar seconds
.Oo
.Sm off
.Fl v
.Op Cm + | -
.Ar val Op Cm y | m | w | d | H | M | S
.Sm on
.Oc
.Op Cm + Ns Ar output_fmt
.\" Set time with the default input format.
.Nm
.Op Fl jnRu
.Op Fl z Ar output_zone
.Op Fl I Ns Op Ar FMT
.Oo
.Sm off
.Fl v
.Op Cm + | -
.Ar val Op Cm y | m | w | d | H | M | S
.Sm on
.Oc
.Sm off
.Oo Oo Oo Oo Oo
.Ar cc Oc
.Ar yy Oc
.Ar mm Oc
.Ar dd Oc
.Ar HH
.Oc Ar MM Op Cm \&. Ar SS
.Sm on
.Op Cm + Ns Ar output_fmt
.\" Set time with the user-provided input format.
.Nm
.Op Fl jnRu
.Op Fl z Ar output_zone
.Op Fl I Ns Op Ar FMT
.Oo
.Sm off
.Fl v
.Op Cm + | -
.Ar val Op Cm y | m | w | d | H | M | S
.Sm on
.Oc
.Fl f Ar input_fmt
.Ar new_date
.Op Cm + Ns Ar output_fmt
.Sh DESCRIPTION
When invoked without arguments, the
.Nm
utility displays the current date and time.
Otherwise, depending on the options specified,
.Nm
will set the date and time or print it in a user-defined way.
.Pp
The
.Nm
utility displays the date and time read from the kernel clock.
When used to set the date and time,
both the kernel clock and the hardware clock are updated.
.Pp
Only the superuser may set the date,
and if the system securelevel (see
.Xr securelevel 7 )
is greater than 1,
the time may not be changed by more than 1 second.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl f Ar input_fmt
Use
.Ar input_fmt
as the format string to parse the
.Ar new_date
provided rather than using the default
.Sm off
.Oo Oo Oo Oo Oo
.Ar cc Oc
.Ar yy Oc
.Ar mm Oc
.Ar dd Oc
.Ar HH
.Oc Ar MM Op Cm \&. Ar SS
.Sm on
format.
Parsing is done using FreeBSD
.Xr strptime 3 .
.It Fl I Ns Op Ar FMT
Use extended
.St -iso8601
output format.
.Ar FMT
may be omitted, in which case the default is
.Cm date .
Valid
.Ar FMT
values are
.Cm date ,
.Cm hours ,
.Cm minutes ,
.Cm seconds ,
and
.Cm ns
.Pq for nanoseconds .
The date and time is formatted to the specified precision.
When
.Ar FMT
is
.Cm hours
.Po or the more precise
.Cm minutes ,
.Cm seconds ,
or
.Cm ns Pc ,
the extended
.St -iso8601
format includes the timezone offset.
.It Fl j
Do not try to set the date.
This allows you to use the
.Fl f
flag in addition to the
.Cm +
option to convert one date format to another.
Note that any date or time components unspecified by the
.Fl f
format string take their values from the current time.
.It Fl n
Obsolete flag, accepted and ignored for compatibility.
.It Fl R
Use RFC 2822 date and time output format.
This is equivalent to using
.Ql %a, %d %b %Y \&%T %z
as
.Ar output_fmt
while
.Ev LC_TIME
is set to the
.Dq C
locale .
.It Fl r Ar seconds
Print the date and time represented by
.Ar seconds ,
where
.Ar seconds
is the number of seconds since the Unix Epoch
(00:00:00 UTC, January 1, 1970;
see
.Xr time 3 ) ,
and can be specified in decimal, octal, or hex.
.It Fl r Ar filename
Print the date and time of the last modification of
.Ar filename .
.It Fl u
Display or set the date in UTC (Coordinated Universal) time.
By default
.Nm
displays the time in the time zone described by
.Pa /etc/localtime
or the
.Ev TZ
environment variable.
.It Fl z Ar output_zone
Just before printing the time, change to the specified timezone;
see the description of
.Ev TZ
below.
This can be used with
.Fl j
to easily convert time specifications from one zone to another.
.It Xo
.Fl v
.Sm off
.Op Cm + | -
.Ar val Op Cm y | m | w | d | H | M | S
.Sm on
.Xc
Adjust (i.e., take the current date and display the result of the
adjustment; not actually set the date) the second, minute, hour, month
day, week day, month or year according to
.Ar val .
If
.Ar val
is preceded by a plus or minus sign,
the date is adjusted forward or backward according to the remaining string,
otherwise the relevant part of the date is set.
The date can be adjusted as many times as required using these flags.
Flags are processed in the order given.
.Pp
When setting values
(rather than adjusting them),
seconds are in the range 0-59, minutes are in the range 0-59, hours are
in the range 0-23, month days are in the range 1-31, week days are in the
range 0-6 (Sun-Sat),
months are in the range 1-12 (Jan-Dec)
and years are in a limited range depending on the platform.
.Pp
On i386, years are in the range 69-38 representing 1969-2038.
On every other platform, years 0-68 are accepted and represent 2000-2068, and
69-99 are accepted and represent 1969-1999.
In both cases, years between 100 and 1900 (both included) are accepted and
interpreted as relative to 1900 of the Gregorian calendar with a limit of 138 on
i386 and a much higher limit on every other platform.
Years starting at 1901 are also accepted, and are interpreted as absolute years.
.Pp
If
.Ar val
is numeric, one of either
.Cm y ,
.Cm m ,
.Cm w ,
.Cm d ,
.Cm H ,
.Cm M
or
.Cm S
must be used to specify which part of the date is to be adjusted.
.Pp
The week day or month may be specified using a name rather than a
number.
If a name is used with the plus
(or minus)
sign, the date will be put forwards
(or backwards)
to the next
(previous)
date that matches the given week day or month.
This will not adjust the date,
if the given week day or month is the same as the current one.
.Pp
When a date is adjusted to a specific value or in units greater than hours,
daylight savings time considerations are ignored.
Adjustments in units of hours or less honor daylight saving time.
So, assuming the current date is March 26, 0:30 and that the DST adjustment
means that the clock goes forward at 01:00 to 02:00, using
.Fl v No +1H
will adjust the date to March 26, 2:30.
Likewise, if the date is October 29, 0:30 and the DST adjustment means that
the clock goes back at 02:00 to 01:00, using
.Fl v No +3H
will be necessary to reach October 29, 2:30.
.Pp
When the date is adjusted to a specific value that does not actually exist
(for example March 26, 1:30 BST 2000 in the Europe/London timezone),
the date will be silently adjusted forward in units of one hour until it
reaches a valid time.
When the date is adjusted to a specific value that occurs twice
(for example October 29, 1:30 2000),
the resulting timezone will be set so that the date matches the earlier of
the two times.
.Pp
It is not possible to adjust a date to an invalid absolute day, so using
the switches
.Fl v No 31d Fl v No 12m
will simply fail five months of the year.
It is therefore usual to set the month before setting the day; using
.Fl v No 12m Fl v No 31d
always works.
.Pp
Adjusting the date by months is inherently ambiguous because
a month is a unit of variable length depending on the current date.
This kind of date adjustment is applied in the most intuitive way.
First of all,
.Nm
tries to preserve the day of the month.
If it is impossible because the target month is shorter than the present one,
the last day of the target month will be the result.
For example, using
.Fl v No +1m
on May 31 will adjust the date to June 30, while using the same option
on January 30 will result in the date adjusted to the last day of February.
This approach is also believed to make the most sense for shell scripting.
Nevertheless, be aware that going forth and back by the same number of
months may take you to a different date.
.Pp
Refer to the examples below for further details.
.El
.Pp
An operand with a leading plus
.Pq Sq +
sign specifies a user-defined format string
which specifies the format in which to display the date and time.
The format string may contain any of the conversion specifications
described in the
.Xr strftime 3
FreeBSD manual page, as well as any arbitrary text.
.Pp
The following extensions to the regular
.Xr strftime 3
syntax are supported:
.Bl -tag -width "xxxx"
.It Cm \&% Ns Ar n Ns Cm N
Replaced by the
.Ar n Ns
-digit fractional part of the number of seconds since the Unix Epoch.
If
.Ar n
is omitted or zero, a default value of 9 is used, resulting in a
number with nanosecond resolution (hence the choice of the letter
.Sq N
for this conversion).
Note that the underlying clock may not necessarily support nanosecond
resolution.
.It Cm \&%-N
As above, but automatically choose the precision based on the reported
resolution of the underlying clock.
If the
.Fl r
option was specified, the default precision of 9 digits is used.
.El
.Pp
A newline
.Pq Ql \en
character is always output after the characters specified by
the format string.
The format string for the default display is
.Dq %+ .
.Pp
If an operand does not have a leading plus sign, it is interpreted as
a value for setting the system's notion of the current date and time.
The canonical representation for setting the date and time is:
.Pp
.Bl -tag -width Ds -compact -offset indent
.It Ar cc
Century
(either 19 or 20)
prepended to the abbreviated year.
.It Ar yy
Year in abbreviated form
(e.g., 89 for 1989, 06 for 2006).
.It Ar mm
Numeric month, a number from 1 to 12.
.It Ar dd
Day, a number from 1 to 31.
.It Ar HH
Hour, a number from 0 to 23.
.It Ar MM
Minutes, a number from 0 to 59.
.It Ar SS
Seconds, a number from 0 to 60
(59 plus a potential leap second).
.El
.Pp
Everything but the minutes is optional.
.Pp
.Nm
understands the time zone definitions from the IANA Time Zone Database,
.Sy tzdata ,
located in
.Pa /usr/share/zoneinfo .
Time changes for Daylight Saving Time, standard time, leap seconds
and leap years are handled automatically.
.Pp
There are two ways to specify the time zone:
.Pp
If the file or symlink
.Pa /etc/localtime
exists, it is interpreted as a time zone definition file, usually in
the directory hierarchy
.Pa /usr/share/zoneinfo ,
which contains the time zone definitions from
.Sy tzdata .
.Pp
If the environment variable
.Ev TZ
is set, its value is interpreted as the name of a time zone definition
file, either an absolute path or a relative path to a time zone
definition in
.Pa /usr/share/zoneinfo .
The
.Ev TZ
variable overrides
.Pa /etc/localtime .
.Pp
If the time zone definition file is invalid,
.Nm
silently reverts to UTC.
.Pp
Previous versions of
.Nm
included the
.Fl d
(set daylight saving time flag) and
.Fl t
(set negative time zone offset) options, but these details are now
handled automatically by
.Sy tzdata .
Modern offsets are positive for time zones ahead of UTC and negative
for time zones behind UTC, but like the obsolete
.Fl t
option, the
.Sy tzdata
files in the subdirectory
.Pa /usr/share/zoneinfo/Etc
still use an older convention where times ahead of UTC are considered
negative.
.Sh ENVIRONMENT
The following environment variable affects the execution of
.Nm :
.Bl -tag -width Ds
.It Ev TZ
The timezone to use when displaying dates.
The normal format is a pathname relative to
.Pa /usr/share/zoneinfo .
For example, the command
.Dq TZ=America/Los_Angeles date
displays the current time in California.
The variable can also specify an absolute path.
See
.Xr environ 7
for more information.
.El
.Sh FILES
.Bl -tag -width /var/log/messages -compact
.It Pa /etc/localtime
Time zone information file for default system time zone.
May be omitted, in which case the default time zone is UTC.
.It Pa /usr/share/zoneinfo
Directory containing time zone information files.
.It Pa /var/log/messages
Record of the user setting the time.
.It Pa /var/log/utx.log
Record of date resets and time changes.
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, 1 if unable to set the date, and 2
if able to set the local date, but unable to set it globally.
.Sh EXAMPLES
The command
.Pp
.Dl "date +%s.%3N"
.Pp
will print the time elapsed since the Unix Epoch with millisecond
precision.
.Pp
The command:
.Pp
.Dl "date ""+DATE: %Y-%m-%d%nTIME: %H:%M:%S"""
.Pp
will display:
.Bd -literal -offset indent
DATE: 1987-11-21
TIME: 13:36:16
.Ed
.Pp
In the Europe/London timezone, the command:
.Pp
.Dl "date -v1m -v+1y"
.Pp
will display:
.Pp
.Dl "Sun Jan  4 04:15:24 GMT 1998"
.Pp
where it is currently
.Ql "Mon Aug  4 04:15:24 BST 1997" .
.Pp
The command:
.Pp
.Dl "date -v1d -v3m -v0y -v-1d"
.Pp
will display the last day of February in the year 2000:
.Pp
.Dl "Tue Feb 29 03:18:00 GMT 2000"
.Pp
So will the command:
.Pp
.Dl "date -v3m -v30d -v0y -v-1m"
.Pp
because there is no such date as the 30th of February.
.Pp
The command:
.Pp
.Dl "date -v1d -v+1m -v-1d -v-fri"
.Pp
will display the last Friday of the month:
.Pp
.Dl "Fri Aug 29 04:31:11 BST 1997"
.Pp
where it is currently
.Ql "Mon Aug  4 04:31:11 BST 1997" .
.Pp
The command:
.Pp
.Dl "date 8506131627"
.Pp
sets the date to
.Ql "June 13, 1985, 4:27 PM" .
.Pp
.Dl "date ""+%Y%m%d%H%M.%S"""
.Pp
may be used on one machine to print out the date
suitable for setting on another.
.Po Use
.Ql "+%m%d%H%M%Y.%S"
with GNU date on
Linux .
.Pc
.Pp
The command:
.Pp
.Dl "date 1432"
.Pp
sets the time to
.Ql "2:32 PM" ,
without modifying the date.
.Pp
The command
.Pp
.Dl "TZ=America/Los_Angeles date -Iseconds -r 1533415339"
.Pp
will display
.Pp
.Dl "2018-08-04T13:42:19-07:00"
.Pp
The command:
.Pp
.Dl "env LC_ALL=C date -j -f ""%a %b %d %T %Z %Y"" ""`env LC_ALL=C date`"" ""+%s"""
.Pp
can be used to parse the output from
.Nm
and express it in Epoch time.
.Pp
Finally the command
.Pp
.Dl "TZ=America/Los_Angeles date -z Europe/Paris -j 0900"
.Pp
will print the time in the
.Dq Europe/Paris
timezone when it is 9:00 in the
.Dq America/Los_Angeles
timezone.
.Sh DIAGNOSTICS
It is invalid to combine the
.Fl I
flag with either
.Fl R
or an output format
.Dq ( + Ns ... )
operand.
If this occurs,
.Nm
prints:
.Ql multiple output formats specified
and exits with status 1.
.Sh SEE ALSO
.Xr locale 1 ,
.Xr clock_gettime 2 ,
.Xr gettimeofday 2 ,
.Xr getutxent 3 ,
.Xr strftime 3 ,
.Xr strptime 3 ,
.Xr tzset 3 ,
.Xr adjkerntz 8 ,
.Xr ntpd 8 ,
.Xr tzsetup 8
.Rs
.%T "TSP: The Time Synchronization Protocol for UNIX 4.3BSD"
.%A R. Gusella
.%A S. Zatti
.Re
.Rs
.%U https://iana.org/time-zones
.%T Time Zone Database
.Re
.Sh STANDARDS
The
.Nm
utility is expected to be compatible with
.St -p1003.2 .
With the exception of the
.Fl u
option, all options are extensions to the standard.
.Pp
The format selected by the
.Fl I
flag is compatible with
.St -iso8601 .
.Pp
The
.Ql \&%N
conversion specification for nanoseconds is a non-standard extension.
It is compatible with GNU date's
.Ql \&%N .
.Sh HISTORY
A
.Nm
command appeared in
.At v1 .
.Pp
A number of options were added and then removed again, including the
.Fl d
(set DST flag) and
.Fl t
(set negative time zone offset).
Time zones are now handled by code bundled with
.Sy tzdata .
.Pp
The
.Fl I
flag was added in
.Fx 12.0 .
.Pp
The
.Ql \&%N
conversion specification was added in
.Fx 14.1 .
Support for the
.Ql \&% Ns Ar n Ns Cm N
and
.Ql \&%-N
variants was added in
.Fx 15.1 .


src.freebsd/coreutils/date/date.c
/*-
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Copyright (c) 1985, 1987, 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#include <sys/param.h>
#include <sys/time.h>
#include <sys/stat.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <inttypes.h>
#include <locale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <utmpx.h>
#include <time_bsd.h>
#include <langinfo.h>

#include "vary.h"

#ifndef	TM_YEAR_BASE
#define	TM_YEAR_BASE	1900
#endif

static void badformat(void);
static void iso8601_usage(const char *) __dead2;
static void multipleformats(void);
static void printdate(const char *);
static void printisodate(struct tm *, long, long);
static void setthetime(const char *, const char *, int, struct timespec *);
static size_t strftime_ns(char * __restrict, size_t, const char * __restrict,
    const struct tm * __restrict, long, long);
static void usage(void) __dead2;

static const struct iso8601_fmt {
	const char *refname;
	const char *format_string;
} iso8601_fmts[] = {
	{ "date", "%Y-%m-%d" },
	{ "hours", "T%H" },
	{ "minutes", ":%M" },
	{ "seconds", ":%S" },
	{ "ns", ",%N" },
};
static const struct iso8601_fmt *iso8601_selected;

static const char *rfc2822_format = "%a, %d %b %Y %T %z";

int
main(int argc, char *argv[])
{
	struct timespec ts = { 0, 0 }, tres = { 0, 1 };
	int ch, rflag;
	bool Iflag, jflag, Rflag;
	const char *format;
	char buf[1024];
	char *end, *fmt, *outzone = NULL;
	struct vary *v;
	const struct vary *badv;
	struct tm *lt;
	struct stat sb;
	size_t i;
	intmax_t number;

	v = NULL;
	fmt = NULL;
	(void) setlocale(LC_TIME, "");
	rflag = 0;
	Iflag = jflag = Rflag = 0;
	while ((ch = getopt(argc, argv, "f:I::jnRr:uv:z:")) != -1)
		switch((char)ch) {
		case 'f':
			fmt = optarg;
			break;
		case 'I':
			if (Rflag)
				multipleformats();
			Iflag = 1;
			if (optarg == NULL) {
				iso8601_selected = iso8601_fmts;
				break;
			}
			for (i = 0; i < nitems(iso8601_fmts); i++)
				if (strcmp(optarg, iso8601_fmts[i].refname) == 0)
					break;
			if (i == nitems(iso8601_fmts))
				iso8601_usage(optarg);

			iso8601_selected = &iso8601_fmts[i];
			break;
		case 'j':
			jflag = 1;	/* don't set time */
			break;
		case 'n':
			break;
		case 'R':		/* RFC 2822 datetime format */
			if (Iflag)
				multipleformats();
			Rflag = 1;
			break;
		case 'r':		/* user specified seconds */
			rflag = 1;
			number = strtoimax(optarg, &end, 0);
			if (end > optarg && *end == '\0') {
				ts.tv_sec = number;
				ts.tv_nsec = 0;
			} else if (stat(optarg, &sb) == 0) {
				ts.tv_sec = sb.st_mtim.tv_sec;
				ts.tv_nsec = sb.st_mtim.tv_nsec;
			} else {
				usage();
			}
			break;
		case 'u':		/* do everything in UTC */
			(void)setenv("TZ", "UTC0", 1);
			break;
		case 'z':
			outzone = optarg;
			break;
		case 'v':
			v = vary_append(v, optarg);
			break;
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (!rflag) {
		if (clock_gettime(CLOCK_REALTIME, &ts) == -1)
			err(1, "clock_gettime");
		if (clock_getres(CLOCK_REALTIME, &tres) == -1)
			err(1, "clock_getres");
	}

	format = "%+";

	if (Rflag)
		format = rfc2822_format;

	/* allow the operands in any order */
	if (*argv && **argv == '+') {
		if (Iflag)
			multipleformats();
		format = *argv + 1;
		++argv;
	}

	if (*argv) {
		setthetime(fmt, *argv, jflag, &ts);
		++argv;
	} else if (fmt != NULL)
		usage();

	if (*argv && **argv == '+') {
		if (Iflag)
			multipleformats();
		format = *argv + 1;
	}

	if (outzone != NULL && setenv("TZ", outzone, 1) != 0)
		err(1, "setenv(TZ)");
	lt = localtime(&ts.tv_sec);
	if (lt == NULL)
		errx(1, "invalid time");
	badv = vary_apply(v, lt);
	if (badv) {
		fprintf(stderr, "%s: Cannot apply date adjustment\n",
		    badv->arg);
		vary_destroy(v);
		usage();
	}
	vary_destroy(v);

	if (Iflag)
		printisodate(lt, ts.tv_nsec, tres.tv_nsec);

	if (format == rfc2822_format)
		/*
		 * When using RFC 2822 datetime format, don't honor the
		 * locale.
		 */
		setlocale(LC_TIME, "C");


	(void)strftime_ns(buf, sizeof(buf), format, lt,
	    ts.tv_nsec, tres.tv_nsec);
	printdate(buf);
}

static void
printdate(const char *buf)
{
	(void)printf("%s\n", buf);
	if (fflush(stdout))
		err(1, "stdout");
	exit(EXIT_SUCCESS);
}

static void
printisodate(struct tm *lt, long nsec, long res)
{
	const struct iso8601_fmt *it;
	char fmtbuf[64], buf[64], tzbuf[8];

	fmtbuf[0] = 0;
	for (it = iso8601_fmts; it <= iso8601_selected; it++)
		strlcat(fmtbuf, it->format_string, sizeof(fmtbuf));

	(void)strftime_ns(buf, sizeof(buf), fmtbuf, lt, nsec, res);

	if (iso8601_selected > iso8601_fmts) {
		(void)strftime_ns(tzbuf, sizeof(tzbuf), "%z", lt, nsec, res);
		memmove(&tzbuf[4], &tzbuf[3], 3);
		tzbuf[3] = ':';
		strlcat(buf, tzbuf, sizeof(buf));
	}

	printdate(buf);
}

#define	ATOI2(s)	((s) += 2, ((s)[-2] - '0') * 10 + ((s)[-1] - '0'))

static void
setthetime(const char *fmt, const char *p, int jflag, struct timespec *ts)
{
	struct utmpx utx;
	struct tm *lt;
	const char *dot, *t;
	struct timeval tv;
	int century;

	lt = localtime(&ts->tv_sec);
	if (lt == NULL)
		errx(1, "invalid time");
	lt->tm_isdst = -1;		/* divine correct DST */

	if (fmt != NULL) {
		t = strptime_bsd(p, fmt, lt);
		if (t == NULL) {
			fprintf(stderr, "Failed conversion of ``%s''"
				" using format ``%s''\n", p, fmt);
			badformat();
		} else if (*t != '\0')
			fprintf(stderr, "Warning: Ignoring %ld extraneous"
				" characters in date string (%s)\n",
				(long) strlen(t), t);
	} else {
		for (t = p, dot = NULL; *t; ++t) {
			if (isdigit(*t))
				continue;
			if (*t == '.' && dot == NULL) {
				dot = t;
				continue;
			}
			badformat();
		}

		if (dot != NULL) {			/* .ss */
			dot++; /* *dot++ = '\0'; */
			if (strlen(dot) != 2)
				badformat();
			lt->tm_sec = ATOI2(dot);
			if (lt->tm_sec > 61)
				badformat();
		} else
			lt->tm_sec = 0;

		century = 0;
		/* if p has a ".ss" field then let's pretend it's not there */
		switch (strlen(p) - ((dot != NULL) ? 3 : 0)) {
		case 12:				/* cc */
			lt->tm_year = ATOI2(p) * 100 - TM_YEAR_BASE;
			century = 1;
			/* FALLTHROUGH */
		case 10:				/* yy */
			if (century)
				lt->tm_year += ATOI2(p);
			else {
				lt->tm_year = ATOI2(p);
				if (lt->tm_year < 69)	/* hack for 2000 ;-} */
					lt->tm_year += 2000 - TM_YEAR_BASE;
				else
					lt->tm_year += 1900 - TM_YEAR_BASE;
			}
			/* FALLTHROUGH */
		case 8:					/* mm */
			lt->tm_mon = ATOI2(p);
			if (lt->tm_mon > 12)
				badformat();
			--lt->tm_mon;		/* time struct is 0 - 11 */
			/* FALLTHROUGH */
		case 6:					/* dd */
			lt->tm_mday = ATOI2(p);
			if (lt->tm_mday > 31)
				badformat();
			/* FALLTHROUGH */
		case 4:					/* HH */
			lt->tm_hour = ATOI2(p);
			if (lt->tm_hour > 23)
				badformat();
			/* FALLTHROUGH */
		case 2:					/* MM */
			lt->tm_min = ATOI2(p);
			if (lt->tm_min > 59)
				badformat();
			break;
		default:
			badformat();
		}
	}

	/* convert broken-down time to GMT clock time */
	lt->tm_yday = -1;
	ts->tv_sec = mktime(lt);
	if (lt->tm_yday == -1)
		errx(1, "nonexistent time");
	ts->tv_nsec = 0;

	if (!jflag) {
		utx.ut_type = OLD_TIME;
		memset(utx.ut_id, 0, sizeof(utx.ut_id));
		(void)gettimeofday(&tv, NULL);
		utx.ut_tv.tv_sec = tv.tv_sec;
		utx.ut_tv.tv_usec = tv.tv_usec;
		pututxline(&utx);
		if (clock_settime(CLOCK_REALTIME, ts) != 0)
			err(1, "clock_settime");
		utx.ut_type = NEW_TIME;
		(void)gettimeofday(&tv, NULL);
		utx.ut_tv.tv_sec = tv.tv_sec;
		utx.ut_tv.tv_usec = tv.tv_usec;
		pututxline(&utx);

		if ((p = getlogin()) == NULL)
			p = "???";
		syslog(LOG_AUTH | LOG_NOTICE, "date set by %s", p);
	}
}

/*
 * The strftime_ns function is a wrapper around strftime(3), which adds support
 * for features absent from strftime(3). Currently, the only extra feature is
 * support for %N, the nanosecond conversion specification.
 *
 * The functions scans the format string for the non-standard conversion
 * specifications and replaces them with the date and time values before
 * passing the format string to strftime(3). The handling of the non-standard
 * conversion specifications happens before the call to strftime(3) to handle
 * cases like "%%N" correctly ("%%N" should yield "%N" instead of nanoseconds).
 */
static size_t
strftime_ns(char * __restrict s, size_t maxsize, const char * __restrict format,
    const struct tm * __restrict t, long nsec, long res)
{
	size_t ret;
	char *newformat;
	char *oldformat;
	const char *prefix;
	const char *suffix;
	const char *tok;
	long number;
	int i, len, prefixlen, width, zeroes;
	bool seen_percent, seen_dash, seen_width;

	seen_percent = false;
	if ((newformat = strdup(format)) == NULL)
		err(1, "strdup");
	tok = newformat;
	for (tok = newformat; *tok != '\0'; tok++) {
		switch (*tok) {
		case '%':
			/*
			 * If the previous token was a percent sign,
			 * then there are two percent tokens in a row.
			 */
			if (seen_percent) {
				seen_percent = false;
			} else {
				seen_percent = true;
				seen_dash = seen_width = false;
				prefixlen = tok - newformat;
				width = 0;
			}
			break;
		case 'N':
			if (!seen_percent)
				break;
			oldformat = newformat;
			prefix = oldformat;
			suffix = tok + 1;
			/*
			 * Prepare the number we are about to print.  If
			 * the requested width is less than 9, we need to
			 * cut off the least significant digits.  If it is
			 * more than 9, we will have to append zeroes.
			 */
			if (seen_dash) {
				/*
				 * Calculate number of singificant digits
				 * based on res which is the clock's
				 * resolution in nanoseconds.
				 */
				for (width = 9, number = res;
				     width > 0 && number > 0;
				     width--, number /= 10)
					/* nothing */;
			}
			number = nsec;
			zeroes = 0;
			if (width == 0) {
				width = 9;
			} else if (width > 9) {
				zeroes = width - 9;
				width = 9;
			} else {
				for (i = 0; i < 9 - width; i++)
					number /= 10;
			}
			/*
			 * Construct a new format string from the prefix
			 * (i.e., the part of the old format from its
			 * beginning to the currently handled "%N"
			 * conversion specification), the nanoseconds, and
			 * the suffix (i.e., the part of the old format
			 * from the next token to the end).
			 */
			asprintf(&newformat, "%.*s%.*ld%.*d%n%s", prefixlen,
			    prefix, width, number, zeroes, 0, &len, suffix);
			if (newformat == NULL)
				err(1, "asprintf");
			free(oldformat);
			tok = newformat + len - 1;
			seen_percent = false;
			break;
		case '-':
			if (seen_percent) {
				if (seen_dash || seen_width) {
					seen_percent = false;
					break;
				}
				seen_dash = true;
			}
			break;
		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			if (seen_percent) {
				if (seen_dash) {
					seen_percent = false;
					break;
				}
				width = width * 10 + *tok - '0';
				seen_width = true;
			}
			break;
		default:
			seen_percent = false;
			break;
		}
	}

	ret = strftime_bsd(s, maxsize, newformat, t);
	free(newformat);
	return (ret);
}

static void
badformat(void)
{
	warnx("illegal time format");
	usage();
}

static void
iso8601_usage(const char *badarg)
{
	errx(1, "invalid argument '%s' for -I", badarg);
}

static void
multipleformats(void)
{
	errx(1, "multiple output formats specified");
}

static void
usage(void)
{
	(void)fprintf(stderr, "%s\n%s\n%s\n",
	    "usage: date [-jnRu] [-I[date|hours|minutes|seconds|ns]] [-f input_fmt]",
	    "            "
	    "[ -z output_zone ] [-r filename|seconds] [-v[+|-]val[y|m|w|d|H|M|S]]",
	    "            "
	    "[[[[[[cc]yy]mm]dd]HH]MM[.SS] | new_date] [+output_fmt]"
	    );
	exit(1);
}
